parser/README.md

```diff
@@ -0,0 +1,134 @@
+
+```go
+func Parse(sql string) (acceptable bool, err error) {
+    // Where SQL is acceptable by the original engine, no error occurred.
+    err := sql_engine.Parse(sql)
+    if err == nil {
+        return true, ni
+    }
+    
+    // Error message from a parser should contain error position.
+    pos := parseErrorPosition(err) 
+    leftPart = sql[:pos]
+    rightPart = sql[pos:]
+
+    errLeft := sql_engine.Parse(leftPart)
+    // In this case, the SQL is not acceptable due to the syntax error
+    if err != nil {
+       return false, err 
+    }
+
+    // If leftPart is acceptable, it is a legitimate  SELECT statement. We then try right part with SQLFlow parser. 
+    errRight := SQLFlow.Parse(rightPart)
+    if err != nil {
+        return false, errRight 
+    }
+
+    // The left part is a SELECT and the right part is TRAIN or PREDICT.
+    return false, nil 
+}
+```
+
+This procedure has two assumptions: (1) a SQL engine has a parser, denoted by sql_engine.Parse, and (2) the parser reports the position of syntax errors if there is any.  Both assumptions are reasonable. However, a key question here is how to call parsers of various SQL engines. For open source SQL engines, this doesn't seem an issue. Here we list parsers used by some well-known engines:
+
+| SQL Engine | Parser  |
+|------------|---------|
+| Beam       | Calcite |
+| Flink      | Calcite |
+| Storm      | Calcite |
+| Hive       | Hplsql.g4 |
+| MySQL      | sql_parse.{h,cc} |
+
+It is notable that MySQL doesn't provide sufficient documentation on how to call its parser, so we call a Go implementation of MySQL's SQL parser by PingCap https://github.com/pingcap/parser instead. For proprietary engines, we have a plan to ask the licensing of their parsers.
+
+
+## The External Parser Abstraction
+
+The above example program calls `sql_engine.Parse` twice before calling `SQLFlow.Parse`.  In practice, because Calcite parser is in Java, SQLFlow is a Go program, to enable SQLFlow calling Calcite parser, we have to wrap Calcite parser up into a gRPC server.  It is time-consuming to make an RPC call, so we pack the two calls to `sql_engine.Parse` into one:
+
+```protobuf
+message CalciteParserRequest {
+  string query = 1;
+}
+
+message CalciteParserReply {
+  string sql = 1;
+  string extension = 2;
+  string error = 3;
+}
+
+service CalciteParser {
+  rpc Parse (CalciteParserRequest) returns (CalciteParserReply) {}
+}
+```
+
+Some external parsers are in Go and don't need RPC.  For example, TiDB parser is in Go, and HiveQL parser is a set of AntLR grammar rules, and AntLR can generate Go code.  However, we would unify the interface to external parsers:
+
+```go
+func external_parser(kind, sql string) (idx int, err error) {
+   switch(kind) {
+   case 'calcite':
+       r := grpc.Call("CalciteParser.Parse", CalciteParserRequest{sql})
+       return r.GetIndex(), r.GetError()
+   case 'mysql':
```

Comment:
should be 'tidb'?

Reply:
It is MySQL. We used TiDB parser for MySQL.